pkg/integrations/v2/controller.go

```diff
@@ -0,0 +1,521 @@
+// integrations.
+func newController(l log.Logger, cfg controllerConfig, globals Globals) (*controller, error) {
+	c := &controller{
+		logger:             l,
+		reloadIntegrations: make(chan struct{}, 1),
+	}
+	if err := c.UpdateController(cfg, globals); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// run starts the controller and blocks until ctx is canceled.
+func (c *controller) run(ctx context.Context) {
+	defer func() {
+		level.Debug(c.logger).Log("msg", "stopping all integrations")
+
+		c.mut.Lock()
+		defer c.mut.Unlock()
+
+		for _, exist := range c.integrations {
+			exist.Stop()
+		}
+	}()
+
+	var currentIntegrations []*controlledIntegration
+
+	updateIntegrations := func() {
+		// Lock the mutex to prevent another set of integrations from being
+		// loaded in.
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		level.Debug(c.logger).Log("msg", "updating running integrations", "prev_count", len(currentIntegrations), "new_count", len(c.integrations))
+
+		newIntegrations := c.integrations
+
+		// Shut down all old integrations. If the integration exists in
+		// newIntegrations but has a different gen number, then there's a new
+		// instance to launch.
+		for _, exist := range currentIntegrations {
+			var found bool
+			for _, current := range newIntegrations {
+				if exist.id == current.id && current.gen == exist.gen {
+					found = true
+					break
+				}
+			}
+			if !found {
+				exist.Stop()
+			}
+		}
+
+		var wg sync.WaitGroup
+		wg.Add(len(newIntegrations))
+
+		// Now all integrations can be launched.
+		for _, current := range newIntegrations {
+			go func(current *controlledIntegration) {
+				err := current.Run(ctx, func() { wg.Done() })
+				if err != nil && !errors.Is(err, errIntegrationRunning) {
+					level.Warn(c.logger).Log("msg", "integration exited with error", "instance", current.id, "err", err)
+				}
+			}(current)
+		}
+
+		// Wait for all integration goroutines to have been scheduled at least once.
+		wg.Wait()
+
+		// Finally, store the current list of contolled integrations.
+		currentIntegrations = newIntegrations
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			level.Debug(c.logger).Log("msg", "controller exiting")
+			return
+		case <-c.reloadIntegrations:
+			updateIntegrations()
+			if c.onUpdateDone != nil {
+				c.onUpdateDone()
+			}
+		}
+	}
+}
+
+// controlledIntegration is a running Integration.
+// A running integration is identified uniquely by its id and gen.
+type controlledIntegration struct {
+	id  integrationID
+	gen uint64
+
+	i Integration
+	c Config // Config that generated i. Used for changing to see if a config changed.
+
+	running uint64 // running must only be used atomically
+
+	mut  sync.Mutex
+	stop context.CancelFunc
+}
+
+func (ci *controlledIntegration) Running() bool {
+	return atomic.LoadUint64(&ci.running) == 1
+}
+
+func (ci *controlledIntegration) Run(ctx context.Context, started func()) error {
+	swapped := atomic.CompareAndSwapUint64(&ci.running, 0, 1)
```

Comment:
I'm not opposed to this, though could you elaborate on why? We do return it back to `0` when it stops running, so `alreadyStarted` might be misleading.

Reply:
It's a nit/style concern so not a big deal, but in my mind `swapped` refers to what happened but not why. Since `atomic.Compare...` is a general use function I have to look at the code for a few seconds to realize what is going on.